pkg/logql/engine.go

```diff
@@ -297,19 +297,28 @@ func PopulateMatrixFromScalar(data promql.Scalar, params LiteralParams) promql.M
 	return promql.Matrix{series}
 }
 
-func readStreams(i iter.EntryIterator, size uint32) (Streams, error) {
+func readStreams(i iter.EntryIterator, size uint32, step time.Duration, dir logproto.Direction) (Streams, error) {
 	streams := map[string]*logproto.Stream{}
 	respSize := uint32(0)
-	for ; respSize < size && i.Next(); respSize++ {
+	// lastEntry should be a really old time so that the first comparison is always true, we use a negative
+	// value here because many unit tests start at time.Unix(0,0)
+	lastEntry := time.Unix(-100, 0)
+	for respSize < size && i.Next() {
 		labels, entry := i.Labels(), i.Entry()
-		stream, ok := streams[labels]
-		if !ok {
-			stream = &logproto.Stream{
-				Labels: labels,
+		if step == 0 || lastEntry.Unix() < 0 ||
+			(dir == logproto.FORWARD && (i.Entry().Timestamp.Equal(lastEntry.Add(step)) || i.Entry().Timestamp.After(lastEntry.Add(step)))) ||
+			(dir == logproto.BACKWARD && (i.Entry().Timestamp.Equal(lastEntry.Add(-step)) || i.Entry().Timestamp.Before(lastEntry.Add(-step)))) {
```

Comment:
I'm struggling to read this conditional a bit because it's so long. Could anything be done to improve it?

This would help my reading of it, but you may have a better idea:

```go
// ... 
if step == 0 || lastEntry.Unix() < 0 {
  if dir == logproto.FORWARD && !timestampEqualOrAfter(i.Entry().Timestamp, lastEntry.Add(step)) {
     continue 
  } else if dir = logproto.BACKWARD && !timestampEqualOrBefore(i.Entry().Timestamp, lastEntry.Add(-step)) {
     continue 
  } 
  // ...
} 
```